mchf-eclipse/drivers/audio/psk.c

```diff
@@ -9,10 +9,63 @@
 
 // Common
 #include "psk.h"
-#include "ui_driver.h"
-#include "radio_management.h"
+#include <math.h>
 #include <stdlib.h>
+#include "softdds.h"
 #include "uhsdr_digi_buffer.h"
+#include "ui_driver.h" // only necessary because of UiDriver_TextMsgPutChar
```

Comment:
Hi Danilo, I thought about it, we can remove it if switch UiDriver_TextMsgPutChar() to DigiModes_TxBufferPutChar( ... , BPSK );
And the same in RTTY. 
Theoretically should work....

Reply:
Just to say something about the current design and code:

The TxBuffer is a buffer serves as input for digital modes which require textual input.
This buffer is filled by one or more of the supported input methods (CW or Keyboard or Keyer). And read by a single consumer in an interrupt (by the active TX digi mode).
 
The UiDriver_TxtMsg buffer serves a input buffer for the console output for single producer of text in an interrupt, the output is in the main program part consumer.

The code for put and get  is safe to be used concurrently but not reentrant, i.e. while put and get may interrupt each other, it is not permitted to call put, while interrupt another put (same applies to get).  The code is non-blocking and require not explicit interrupt disabling or any other kind of synchronization or mutexes on critical sections.

For this reason these buffers shoud not be mixed. 

It is has not been the intention of these buffers to provide things like editable text (before it is sent) and to actually handle the display of both data on screen. That would be the purpose of a separate functionality on the UI side and should be decoupled from my point of view.

I agree that the UiDriver_TxtMsg buffer should not be part of the ui_driver.c and could be moved to the other buffer code serving the opposite direction or in a file nearby.

Reply:
Thank you for replay.

> Just to say something about the current design and code:
> 
> The TxBuffer is a buffer serves as input for digital modes which require textual input.
> This buffer is filled by one or more of the supported input methods (CW or Keyboard or Keyer). And read by a single consumer in an interrupt (by the active TX digi mode).
> 

Not only, Recording macros is also reading from it, not from interrupt as far as I know. 

> The UiDriver_TxtMsg buffer serves a input buffer for the console output for single producer of text in an interrupt, the output is in the main program part consumer.
> 

And here agoain, we have UI which produses ont from interrupt.
> The code for put and get is safe to be used concurrently but not reentrant, i.e. while put and get may interrupt each other, it is not permitted to call put, while interrupt another put (same applies to get). The code is non-blocking and require not explicit interrupt disabling or any other kind of synchronization or mutexes on critical sections.
> 
> For this reason these buffers shoud not be mixed.
> 

I see your main point, dont warry I keep it in mind.